form/form.go

```diff
@@ -0,0 +1,279 @@
+	// should be an empty string. It's used to workaround the fact that an
+	// empty string is a string's zero value and wouldn't normally be encoded.
+	Empty bool
+
+	// Invert indicates that a boolean field's value should be inverted. False
+	// is the zero value for a boolean so it's convention in the library to
+	// specify a `No*` field to allow a false to be passed to the API. These
+	// fields should be annotated with `invert`.
+	Invert bool
+
+	// Zero indicates a field that's specifically defined to workaround the
+	// fact because 0 is the "zero value" of all int/float types, we can't
+	// properly encode an explicit 0. It indicates that an explicit zero should
+	// be sent.
+	Zero bool
+}
+
+func AppendTo(values *RequestValues, i interface{}) {
+	reflectValue(values, reflect.ValueOf(i), nil, nil)
+}
+
+func formatName(names []string) string {
+	if len(names) < 1 {
+		panic("Not allowed 0-length names slice")
+	}
+
+	fullName := names[0]
+	for i := 1; i < len(names); i++ {
+		fullName += "[" + names[i] + "]"
+	}
+	return fullName
+}
+
+func reflectValue(values *RequestValues, val reflect.Value, names []string, options *formOptions) {
+	t := val.Type()
+
+	// Also do nothing if this is the type's zero value
+	if t.Comparable() && val.Interface() == reflect.Zero(t).Interface() {
+		return
+	}
+
+	switch val.Kind() {
+	case reflect.Array, reflect.Slice:
+		// formatName automatically adds square brackets, so just pass an empty
+		// string into the breadcrumb trail
+		arrNames := append(names, "")
+
+		for i := 0; i < val.Len(); i++ {
+			// The one exception to the above is when options have requested
+			// that this array/slice be indexed. In that case we produce a hash
+			// keyed with integers which the Stripe API knows how to interpret.
+			if options != nil && options.IndexedArray {
+				arrNames = append(names, strconv.Itoa(i))
+			}
+
+			reflectValue(values, val.Index(i), arrNames, nil)
+		}
+
+	case reflect.Bool:
+		if options != nil {
+			if val.Bool() {
+				switch {
+				case options.Empty:
+					values.Add(formatName(names), "")
+				case options.Invert:
+					values.Add(formatName(names), strconv.FormatBool(false))
+				case options.Zero:
+					values.Add(formatName(names), "0")
+				}
+			}
+		} else {
+			values.Add(formatName(names), strconv.FormatBool(val.Bool()))
+		}
+
+	case reflect.Float32:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 32))
+
+	case reflect.Float64:
+		values.Add(formatName(names), strconv.FormatFloat(val.Float(), 'f', 4, 64))
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		values.Add(formatName(names), strconv.FormatInt(val.Int(), 10))
+
+	case reflect.Interface:
+		if val.IsNil() {
+			return
+		}
+		reflectValue(values, val.Elem(), names, nil)
+
+	case reflect.Map:
+		for _, keyVal := range val.MapKeys() {
+			if keyVal.Kind() != reflect.String {
+				panic("Don't support serializing maps with non-string keys")
```

Comment:
I don't think there's anything in our library besides string-keyed maps, but you could have a `map[int]string` or whatever. I'm trying to help ensure that people just don't try to do this by panicking so that we can catch the problem early :)

I could see the argument for returning an `error` though, which would be fairly conventional. I think we could refactor that over, but I'm basically following the convention set by today's `AppendTo` functions which also don't return errors.